libraries/Microsoft.Bot.Connector/Authentication/BotFrameworkAuthentication.cs

```diff
@@ -79,6 +82,38 @@ public virtual Task<ClaimsIdentity> AuthenticateChannelRequestAsync(string authH
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Creates a <see cref="StreamingConnection"/> that uses web sockets.
+        /// </summary>
+        /// <param name="httpContext"><see cref="HttpContext"/> instance on which to accept the web socket.</param>
+        /// <param name="logger">Logger implementation for tracing and debugging information.</param>
+        /// <returns><see cref="StreamingConnection"/> that uses web socket.</returns>
+        public virtual Task<StreamingConnection> CreateWebSocketConnectionAsync(HttpContext httpContext, ILogger logger)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            return Task.FromResult<StreamingConnection>(new WebSocketStreamingConnection(httpContext, logger));
+        }
+
+        /// <summary>
+        /// Creates a <see cref="StreamingConnection"/> that uses named pipes.
+        /// </summary>
+        /// <param name="pipeName">The name of the named pipe.</param>
+        /// <param name="logger">Logger implementation for tracing and debugging information.</param>
+        /// <returns>A <see cref="StreamingConnection"/> that uses named pipes.</returns>
+        public virtual StreamingConnection CreateNamedPipeConnection(string pipeName, ILogger logger)
+        {
+            if (string.IsNullOrWhiteSpace(pipeName))
+            {
+                throw new ArgumentNullException(nameof(pipeName));
+            }
+
+            return new LegacyStreamingConnection(pipeName, logger);
+        }
```

Comment:
Could you explain the rationale for adding these methods? To me this appears to be a muddling of the SDK layers, so I'm curious about the factors driving this change.

Reply:
There were a couple of reasons for these methods:
1) We introduced the new pipelines based streaming protocol, which will live side-by-side to the legacy streaming protocol for some time before we deprecate the legacy implementation. This method encapsulates which `StreamingConnection` gets constructed (legacy/new). The bot author can control this from IConfiguration.
2) In order to unit test the streaming code path, we need a way to inject a mock `StreamingConnection`, which could be possible by overriding these methods.

Reply:
I dont think we need the pipes one. The real need here is to implement a Strategy pattern where there are 2 connection strategies. The `LegacyConnection` which lives in `Microsoft.Bot.Builder` and the new `WebSocketConnection` which lives in `Microsoft.Bot.Connector.Streaming`. One way is to implement the strategy pattern here through the `CreateWEbSocketConnection` method, another option for example is to implement it in the `CloudAdapter` by adding `IConfiguration` to all constructors and selecting strategy based on the `IConfiguration`.

Reply:
Another way to avoid this layering issue is add an intermediate authentication class higher up which knows about StreamingConnections and have the adapter use that one to build the connection

Reply:
Removed these methods and added a protected virtual method to CloudAdapter that creates web socket connection. If bot authors need to fall back to legacy, they would need to override this method to create a `LegacyStreamingConnection`.